--- a/modules/openssl.cpp
+++ b/modules/openssl.cpp
@@ -30,6 +30,7 @@
 
 #ifndef OPENSSL_NO_AES
 #include <openssl/aes.h>
+#include <openssl/modes.h>
 #endif
 
 #ifndef OPENSSL_NO_DES
@@ -644,14 +645,15 @@ bool AesCtrCipher::encrypt(void* outData
 	inpData = outData;
     unsigned int num = 0;
     unsigned char eCountBuf[AES_BLOCK_SIZE];
-    AES_ctr128_encrypt(
+    CRYPTO_ctr128_encrypt(
 	(const unsigned char*)inpData,
 	(unsigned char*)outData,
 	len,
 	m_key,
 	m_initVector,
 	eCountBuf,
-	&num);
+	&num,
+	(block128_f)AES_encrypt);
     return true;
 }
 
